homeassistant/components/modem_callerid/__init__.py

```diff
@@ -1 +1,72 @@
-"""The modem_callerid component."""
+"""The Modem Caller ID integration."""
+import asyncio
+import logging
+
+from phone_modem import PhoneModem, exceptions
+import serial
+
+from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
+from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
+from homeassistant.const import CONF_DEVICE
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+
+from .const import DATA_KEY_API, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = [SENSOR_DOMAIN]
+
+
+async def async_setup(hass: HomeAssistant, config):
+    """Set up the Modem Caller ID component."""
+    hass.data[DOMAIN] = {}
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config.get(DOMAIN)
+        )
+    )
+
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up Modem Caller ID from a config entry."""
+    device = entry.data[CONF_DEVICE]
+    try:
+        api = PhoneModem(device)
+    except (
+        FileNotFoundError,
+        exceptions.SerialError,
+        serial.SerialException,
+        serial.serialutil.SerialException,
+    ) as ex:
+        _LOGGER.error("Unable to open port %s", device)
+        raise ConfigEntryNotReady from ex
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        DATA_KEY_API: api,
+    }
+
```

Comment:
Register an event listener wrapped in `entry.async_on_unload` to stop the `phone-modem` thread on the stop event.

Reply:
See `sensor.py`

Reply:
I am getting `'ConfigEntry' object has no attribute 'async_on_unload'`, tried both in `__init__.py` and `sensor.py`.

Reply:
This was added in dev https://github.com/home-assistant/core/blob/dev/homeassistant/config_entries.py#L530 recently.  You may need to rebase.

https://developers.home-assistant.io/docs/development_catching_up/